File: src/main/java/dev/vrba/slack/tenor/bot/controller/SlackCommandController.java
package dev.vrba.slack.tenor.bot.controller;

import dev.vrba.slack.tenor.bot.commands.TenorSlashCommand;
import dev.vrba.slack.tenor.bot.commands.TenorSlashCommandPayload;
import io.micronaut.core.annotation.NonNull;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.annotation.Body;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Post;
import lombok.AllArgsConstructor;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

import java.util.Map;

@Controller
@AllArgsConstructor
public class SlackCommandController {

    @NonNull
    private final TenorSlashCommand tenor;

    @Post("/tenor")
    @NonNull
    public Mono<HttpResponse<Void>> handleTenorCommand(@NonNull @Body Map<String, String> request) {
        final var payload = new TenorSlashCommandPayload(
            request.get("query"),
            request.get("trigger_id"),
            request.get("response_url")
        );

        tenor.run(payload)
            .subscribeOn(Schedulers.boundedElastic())
            .subscribe();

        return Mono.just(HttpResponse.noContent());
    }
}

File: src/main/kotlin/dev/vrba/slack/controller/SlackCommandController.kt
package dev.vrba.slack.controller

import dev.vrba.slack.commands.TenorSlashCommand
import dev.vrba.slack.commands.TenorSlashCommandPayload
import io.micronaut.http.HttpResponse
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Body
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Post
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch

@Controller("/api/command")
class SlackCommandController(
    private val tenor: TenorSlashCommand,
) {
    @Post(
        value = "/tenor",
        consumes = [MediaType.APPLICATION_FORM_URLENCODED],
        produces = [MediaType.APPLICATION_JSON],
    )
    suspend fun handleTenorCommand(
        @Body body: Map<String, String>,
    ): HttpResponse<String> {
        val payload =
            TenorSlashCommandPayload(
                query = body["text"].orEmpty(),
                trigger = body["trigger_id"].orEmpty(),
                response = body["response_url"].orEmpty(),
            )

        coroutineScope {
            launch(Dispatchers.IO) {
                tenor.run(payload)
            }
        }

        return HttpResponse.noContent()
    }
}
